Resolves #81

CinquilCinquil opened the pull request:
#81

CinquilCinquil committed:
chore(issue-81): Create a base class for saving and loading

#81

CinquilCinquil commented:
I saw that there was already a [savable](https://github.com/UFRN-URNAI/urnai-tools/blob/main/urnai/base/savable.py) class on the previous version of URNAI, so i tested it, fixed some bugs and i think its ok to use now.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,225 @@
+import tempfile
+from multiprocessing import Process
+
+#import time
+#from urnai.utils.reporter import Reporter as rp
```
Remove this.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,225 @@
+#from urnai.utils.reporter import Reporter as rp
+
+
+class SavableAttr:
+    def __init__(self, value):
+        self.value = value
```
I think we can remove this and use a `Dict` instead.

CinquilCinquil replied:
I noticed that this class wasn't being used for anything, so i just removed it, ok?

alvarofpp replied:
Yes.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,225 @@
+
+    def __init__(self, threaded_saving=False):
+        self.threaded_saving = threaded_saving
+        self.pickle_black_list = []
```
`blacklist` is a racist term. Use `blocklist` instead.

alvarofpp replied:
`pickle` restricts the type of file that will be saved. I would use a more general term, as we intend to use several types of files.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,225 @@
+        """
+        pass
+
+    def get_full_persistance_pickle_path(self, persist_path):
+        """This method returns the default persistance pickle path."""
+        return (persist_path + os.path.sep + self.get_default_save_stamp()
+		+ '.pkl')
+
+    def get_full_persistance_tensorflow_path(self, persist_path):
+        """This method returns the default persistance tensorflow path."""
+        return (persist_path + os.path.sep + self.get_default_save_stamp()
+		+ 'tensorflow_')
+
+    def get_full_persistance_path(self, persist_path):
+        """This method returns the default persistance path."""
+        return persist_path + os.path.sep + self.get_default_save_stamp()
+
+    def get_full_persistance_pytorch_path(self, persist_path):
+        """This method returns the default persistance pytorch path."""
+        return (persist_path + os.path.sep + self.get_default_save_stamp()
+		+ '.pt')
```
I would remove all these functions, with the exception of `get_full_persistance_path`.

alvarofpp reviewed (changes requested):
I noticed that `load` methods are in the `Savable` class, I think they should be in another class or we could change the name of `Savable` to something that encompasses these functions.

CinquilCinquil committed:
refactor: Adjusted to reviewer's suggestions

. renamed black_list to block_list
. renamed pickle_block_list to attr_block_list
. removed SavableAttr class
. left only get_full_persistence_path of
  the get_full_x_persistence_path methods

CinquilCinquil commented:
For testing, i woud suggest first creating a class, that inherits from Savable, with some attributes. Then instantiate it, giving random values to the attributes, after that save it and load it to see if it still has the same values as before. I belive all methods of the class are used in this procedure, and that should reveal if any of them are not functional.

CinquilCinquil commented:
Should i make unit tests?

alvarofpp commented:
> For testing, i woud suggest first creating a class, that inherits from Savable, with some attributes. Then instantiate it, giving random values to the attributes, after that save it and load it to see if it still has the same values as before. I belive all methods of the class are used in this procedure, and that should reveal if any of them are not functional.

The Savable class is not abstract, so you should use it directly in the tests. As these are unit tests, you should ideally create tests for each function separately.

alvarofpp commented:
> Should i make unit tests?

Yes, you should.

CinquilCinquil commented:
What about the name Persistence for the class?

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -155,18 +132,18 @@ def get_pickleable_attributes(self):
-                          and a not in self.pickle_black_list
-                          and a != 'pickle_black_list']
+                          and a not in self.attr_block_list
+                          and a != 'attr_block_list']
```
Rename `a` to something more explicit.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -155,18 +132,18 @@ def get_pickleable_attributes(self):
         pickleable attribute, put it
-        in self.pickle_black_list as a
+        in self.attr_block_list as a
         string.
```
You can add more characters to the lines.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -218,8 +195,8 @@ def get_pickleable_dict(self):
 
     def restore_pickleable_attributes(self, dict_to_restore):
```
Rename the function.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -218,8 +195,8 @@ def get_pickleable_dict(self):
         for attr in dict_to_restore:
-            if hasattr(self, 'pickle_black_list'):
-                if attr not in self.pickle_black_list and attr != 'pickle_black_list':
+            if hasattr(self, 'attr_block_list'):
```
`attr_block_list` is started in `__init__`, so I can't imagine a case where it isn't in `self`.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -218,8 +195,8 @@ def get_pickleable_dict(self):
 
     def restore_pickleable_attributes(self, dict_to_restore):
         for attr in dict_to_restore:
```
```suggestion
for key in dict_to_restore:
```

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -218,8 +195,8 @@ def get_pickleable_dict(self):
+                if attr not in self.attr_block_list and attr != 'attr_block_list':
                     setattr(self, attr, dict_to_restore[attr])
             else:
                 setattr(self, attr, dict_to_restore[attr])
```
Avoid `setattr`.

CinquilCinquil replied:
Any suggestions on how? i've found this [page](https://snyk.io/pt-BR/blog/the-dangers-of-setattr-avoiding-mass-assignment/) and it shows mutiple ways, can i choose any of them?

alvarofpp replied:
You can use whichever approach you prefer. If there are only a few attributes, I recommend making it as explicit as possible. If it's a lot of attributes, I think it's better to validate the dictionary schema and then copy it.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -218,8 +195,8 @@ def get_pickleable_dict(self):
-                if attr not in self.pickle_black_list and attr != 'pickle_black_list':
+            if hasattr(self, 'attr_block_list'):
+                if attr not in self.attr_block_list and attr != 'attr_block_list':
                     setattr(self, attr, dict_to_restore[attr])
```
Avoid `setattr`.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -155,18 +132,18 @@ def get_pickleable_attributes(self):
+                          and a != 'attr_block_list']
         pickleable_list = []
 
         for attr in full_attr_list:
```
```suggestion
for key in full_attr_list:
```

alvarofpp commented:
> What about the name Persistence for the class?

It's okay for now.

CinquilCinquil committed:
refactor: Adjusted to some of reviewer's suggestions

#84

CinquilCinquil committed:
refactor: More alterations

#84

CinquilCinquil commented:
Btw, my lint checks keep failing because of the description of an old commit (I put too many characters in a line apparently). Is there any way to fix that?

CinquilCinquil committed:
fix: Replaced attr with key

alvarofpp commented:
> Btw, my lint checks keep failing because of the description of an old commit (I put too many characters in a line apparently). Is there any way to fix that?

You can use interactive rebase (`git rebase -i`) for that.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+        self.threaded_saving = threaded_saving
+        self.attr_block_list = []
+
+    def get_default_save_stamp(self):
```
Most likely, this function will only be used by the class and not by the developer. I would add the prefix `_`.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+        stuff in a separate thread
+        """
+        #rp.report('THREADED Saving {} object...'
+		#.format(self.__class__.__name__), verbosity_lvl=1)
```
Remove comment.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+        rp.report('It took {} seconds to save {} object!'
+		.format(end_time - start_time, self.__class__.__name__),
+        verbosity_lvl=2)
+        '''
```
Remove comment.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+        """
+        #rp.report('Saving {} object...'.format
+		#(self.__class__.__name__), verbosity_lvl=1)
+        #start_time = time.time()
```
Remove comment.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+        """
+        path = self.get_full_persistance_path(persist_path)
+		
+		# creating necessary directories for the file
```
Remove comment.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+        """
+        return self.__class__.__name__ + '_'
+
+    def save_pickle(self, persist_path):
```
It should be a generalized method and then a class for the pickle format that would contain the code for that specific format.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+        """
+        pass
+
+    def load_pickle(self, persist_path):
```
It should be a generalized method and then a class for the pickle format that would contain the code for that specific format.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+
+        return pickleable_list
+
+    def get_pickleable_dict(self):
```
It should be a generalized method and then a class for the pickle format that would contain the code for that specific format.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+        self.load_pickle(loadpath)
+        self.load_extra(loadpath)
+
+    def get_pickleable_attributes(self):
```
It should be a generalized method and then a class for the pickle format that would contain the code for that specific format.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+        verbosity_lvl=2)
+        '''
+
+    def threaded_pickle_save(self, savepath):
```
It should be a generalized method and then a class for the pickle format that would contain the code for that specific format.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+        This method loads a list instance
+        saved by pickle.
+        """
+        # Check if pickle file exists
```
Remove comment.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+        pickle_path = self.get_full_persistance_path(persist_path)
+        exists_pickle = os.path.isfile(pickle_path)
+        
+        # If yes, load it
```
Remove comment.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+            with open(pickle_path, 'rb') as pickle_in:
+                pickle_dict = pickle.load(pickle_in)
+                self.restore_attributes(pickle_dict)
+                '''
+                rp.report(
+                    '**************************************** \n Loaded pickle for '
+                    + self.get_default_save_stamp()
+                    + ' loaded. \n****************************************',
+                2)
+				'''
```
Remove comment.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+    def restore_attributes(self, dict_to_restore):
+        for key in dict_to_restore:
+            if key not in self.attr_block_list and key != 'attr_block_list':
+                setattr(self, key, dict_to_restore[key])
```
Avoid using `setattr`.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+                    continue
+            
+            except ValueError as ve:
+                if 'ctypes objects' in str(ve):
+                    continue
+                else:
+                    raise
```
Remove `else`, change the logic if necessary.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+                    raise
+            
+            except NotImplementedError as nie:
+                if (str(nie) == 
+                'numpy() is only available when eager execution is enabled.'):
+                    continue
+                else:
+                    raise
```
Remove `else`, change the logic if necessary.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+                continue
+            
+            except TypeError as te:
+                if "can't pickle" in str(te):
+                    continue
+                if 'cannot pickle' in str(te):
+                    continue
+                else:
+                    raise
```
Remove `else`, change the logic if necessary.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+		#(self.__class__.__name__), verbosity_lvl=1)
+        #start_time = time.time()
+        
+        if not self.threaded_saving:
+            self.save_pickle(savepath)
+        else:
+            self.threaded_pickle_save(savepath)
```
Remove `else`, change the logic if necessary.

CinquilCinquil replied:
Why?

alvarofpp replied:
This was explained at the last meeting.

alvarofpp commented on urnai/base/savable.py:
```diff
@@ -0,0 +1,196 @@
+        #rp.report('THREADED Saving {} object...'
+		#.format(self.__class__.__name__), verbosity_lvl=1)
+		
+        p = Process(target=self.save_pickle, args=(savepath,))
```
Change the variable name to something more explicit.

CinquilCinquil committed:
refactor: Adjusted to some of reviewer's suggestions

CinquilCinquil committed:
refactor: Changed class and file name to Persistence

#84

CinquilCinquil committed:
feat: Create class PersistencePickle

.Persistence class is now abstract
.Furthermore, pickle related behaviour
was moved to PersistencePickle

CinquilCinquil committed:
feat: Generalize some of Persistence's methods

#84

CinquilCinquil committed:
refactor: Remove unused import

#84

alvarofpp commented on urnai/base/persistence.py:
```diff
@@ -0,0 +1,83 @@
+        """
+        ...
+
+    def load(self, loadpath):
+        self._load(loadpath)
+
+    @abstractmethod
+    def _load(self, savepath):
+        """
+        This method handles logic related
+        to the child class
+        """
+        ...
```
Instead, you can simply set `load` to `abstractmethod`.

```suggestion
    @abstractmethod
    def load(self, loadpath):
        """
        This method handles logic related
        to the child class
        """
        ...
```

alvarofpp commented on urnai/base/persistence.py:
```diff
@@ -0,0 +1,83 @@
+        new_process = Process(target=self._save, args=(savepath,))
+        new_process.start()
+
+    @abstractmethod
+    def _save(self, savepath):
+        """
+        This method handles logic related
+        to the child class
+        """
+        ...
```
To avoid confusion with the `save` method, I would rename this to something like `simple_save`.

alvarofpp commented:
Unit tests are missing.

CinquilCinquil committed:
refactor: Rename _save to _simple_save

#84

CinquilCinquil committed:
test(issue-81): Add unit tests for Persistence

#84

alvarofpp reviewed (changes requested):
The tests for `PersistencePickle` are missing.

CinquilCinquil committed:
test(issue-81): Add unit tests for PersistencePickle

#84

CinquilCinquil commented:
In the `_simple_save` and `load` tests a file is created, but i wish to delete it after the test is done. I tried `os.remove` but it needs extra permission to work. Any suggestions?

alvarofpp commented:
@CinquilCinquil I recommend mocking the functions that create and read the file, that way the file won't be created and won't need to be removed later.

CinquilCinquil commented:
Oh, but the whole purpose of `test_simple_save`, for example, is to check if a file has been created. I'm not familiar with mocking, is  there a way of faking file creation?

alvarofpp commented:
Creating files is something that can result storage costs, and you have to make sure that each test involving `_simple_save` generates a different file so that you don't have the problem of trying to save to the same file or the file not existing because of another test. In short, it's something that can cause more problems than it solves. To test this, I believe that flow/process tests are more suitable than unit tests.

What you can do is mock the function that saves or reads the file and change its execution, in which case you should apply the mock to the `return_value` of the function.

CinquilCinquil committed:
test(issue-81): Mock simple_save and load

#84

alvarofpp commented on urnai/base/persistence_pickle.py:
```diff
@@ -0,0 +1,111 @@
+        Finally the nulled attributes are
+        restored.
+        """
+        path = super().get_full_persistance_path(persist_path)
```
```suggestion
        path = self.get_full_persistance_path(persist_path)
```

alvarofpp commented on urnai/base/persistence_pickle.py:
```diff
@@ -0,0 +1,111 @@
+        This method loads a list instance
+        saved by pickle.
+        """
+        pickle_path = super().get_full_persistance_path(persist_path)
```
```suggestion
        pickle_path = self.get_full_persistance_path(persist_path)
```

alvarofpp commented on urnai/base/persistence_pickle.py:
```diff
@@ -0,0 +1,111 @@
+		If you wish to block one particular pickleable attribute, put it
+        in self.attr_block_list as a string.
+        """
+        if not hasattr(self, 'attr_block_list') or self.attr_block_list is None:
+            self.attr_block_list = []
+
+        full_attr_list = [attr for attr in dir(self) if not attr.startswith('__')
+                          and not callable(getattr(self, attr))
+                          and attr not in self.attr_block_list
+                          and attr != 'attr_block_list'
+                          and attr != 'processes'
+                          and 'abc' not in attr]
```
```suggestion
        if not hasattr(self, 'attr_block_list') or self.attr_block_list is None:
            self.attr_block_list = []

        attr_block_list = self.attr_block_list + ['attr_block_list', 'processes']

        full_attr_list = [attr for attr in dir(self) if not attr.startswith('__')
                          and not callable(getattr(self, attr))
                          and attr not in attr_block_list
                          and 'abc' not in attr]
```

alvarofpp commented on urnai/base/persistence_pickle.py:
```diff
@@ -0,0 +1,111 @@
+            
+            except ValueError as value_error:
+                if 'ctypes objects' not in str(value_error):
+                    raise
```
Raise what?

alvarofpp commented on urnai/base/persistence_pickle.py:
```diff
@@ -0,0 +1,111 @@
+            except TypeError as type_error:
+                if ("can't pickle" not in str(type_error) or
+                 'cannot pickle' not in str(type_error)):
+                    raise
```
Raise what?

alvarofpp commented on urnai/base/persistence_pickle.py:
```diff
@@ -0,0 +1,111 @@
+            except NotImplementedError as notimpl_error:
+                if (str(notimpl_error) != 
+                'numpy() is only available when eager execution is enabled.'):
+                    raise
```
Raise what?

CinquilCinquil committed:
refactor(issue-81): Adjusted to reviewer's suggestions

#84

CinquilCinquil committed:
refactor(issue-81): Adjusted to lint's suggestions

#84

CinquilCinquil committed:
refactor(issue-81): Tests now in "given when then" format

#84
